controllers/statusreport/statusreport_adapter.go

```diff
@@ -86,3 +91,212 @@ func isFeatureEnabled() bool {
 	}
 	return false
 }
+
+// EnsureSnapshotFinishedAllTests is an operation that will ensure that a pipeline Snapshot
+// to the PipelineRun being processed finished and passed all tests for all defined required IntegrationTestScenarios.
+// If the Snapshot doesn't have the freshest state of components, a composite Snapshot will be created instead
+// and the original Snapshot will be marked as Invalid.
+func (a *Adapter) EnsureSnapshotFinishedAllTests() (controller.OperationResult, error) {
+	// Get all required integrationTestScenarios for the Application and then use the Snapshot status annotation
+	// to check if all Integration tests were finished for that Snapshot
+	integrationTestScenarios, err := a.loader.GetRequiredIntegrationTestScenariosForApplication(a.client, a.context, a.application)
```

Comment:
This can be potential race condition, if new required test is added during snapshot testing, this may fetch updated list of required tests but snapshot_adapter may not create new test because controller won't trigger reconciliation without status change.

But this is not new I'd say, this race condition must be there already.

Reply:
I think that race condition could be resolved with a bit of changes on the [current predicate](https://github.com/redhat-appstudio/integration-service/blob/df439eea42b806115b82ff04a0fba9c417721005/controllers/snapshot/snapshot_controller.go#L162), I'll file a follow-up story for that.

Reply:
Opened [STONEINTG-646](https://issues.redhat.com/browse/STONEINTG-646) to track the mentioned issue.